src/Network/NWProtocolStack.cs

```diff
@@ -82,7 +98,18 @@ public void IterateProtocols (Action<NWProtocolOptions> callback)
 		public NWProtocolOptions TransportProtocol {
 			get {
 				var pHandle = nw_protocol_stack_copy_transport_protocol (GetCheckedHandle ());
-				return (pHandle == IntPtr.Zero)? null : new NWProtocolOptions (pHandle, owns: true);
+				if (pHandle == IntPtr.Zero)
+					return null;
+				using (var tempOptions = new NWProtocolOptions (pHandle, owns: false)) 
+				using (var definition = tempOptions.ProtocolDefinition) {
+					if (definition.Equals (NWProtocolDefinition.TcpDefinition)) {
+						return new NWProtocolTcpOptions (pHandle, owns: true);
+					}
+					if (definition.Equals (NWProtocolDefinition.UdpDefinition)) {
+						return new NWProtocolUdpOptions (pHandle, owns: true);
+					} 
+					return null;
```

Comment:
It would be wrong to return `tempOptions` here?

Reply:
I think is the way to go. At least gives something useful.